File: budgeteer/ingest.py
from __future__ import annotations

from pathlib import Path

import pandas as pd

from budgeteer.models import (
    AnyCashFlow,
    Direction,
    Frequency,
    IngestionError,
    LiquidityActual,
    OneOffCashFlow,
    Phase,
    RecurringCashFlow,
)


def load_inputs(
    path: Path,
) -> tuple[list[Phase], list[AnyCashFlow], list[LiquidityActual]]:
    if not path.exists():
        raise IngestionError(f"File not found: {path}")

    phases = _read_phases(path)
    recurring = _read_recurring(path)
    one_offs = _read_one_offs(path)
    actuals = _read_actuals(path)

    return phases, [*recurring, *one_offs], actuals


def _read_sheet(path: Path, sheet_name: str) -> pd.DataFrame:
    try:
        df = pd.read_excel(path, sheet_name=sheet_name, engine="odf")
    except ValueError as e:
        raise IngestionError(f"Sheet '{sheet_name}' not found in {path}") from e
    return df.dropna(how="all")


def _require_columns(df: pd.DataFrame, columns: list[str], sheet_name: str) -> None:
    missing = set(columns) - set(df.columns)
    if missing:
        raise IngestionError(f"Sheet '{sheet_name}': missing columns {sorted(missing)}")


def _to_date(val) -> object:
    if pd.isna(val):
        return None
    return pd.Timestamp(val).date()


def _parse_direction(val: str, item_name: str) -> Direction:
    try:
        return Direction(val.strip().title())
    except (ValueError, AttributeError) as e:
        raise IngestionError(
            f"Cash flow '{item_name}': Direction must be 'Inflow' or 'Outflow', got '{val}'"
        ) from e


def _parse_frequency(val: str, item_name: str) -> Frequency:
    try:
        return Frequency(val.strip().title())
    except (ValueError, AttributeError) as e:
        raise IngestionError(
            f"Cash flow '{item_name}': Frequency must be 'Monthly' or 'Annually', got '{val}'"
        ) from e


def _read_phases(path: Path) -> list[Phase]:
    df = _read_sheet(path, "Phases")
    _require_columns(df, ["Name", "Start_Date", "End_Date"], "Phases")

    phases = []
    for _, row in df.iterrows():
        start = _to_date(row["Start_Date"])
        end = _to_date(row["End_Date"])
        if start is None or end is None:
            raise IngestionError(f"Phase '{row['Name']}': Start_Date and End_Date are required")
        try:
            phases.append(Phase(str(row["Name"]), start, end))
        except ValueError as e:
            raise IngestionError(str(e)) from e

    phases.sort(key=lambda p: p.start_date)
    _validate_no_overlaps(phases)
    return phases


def _validate_no_overlaps(phases: list[Phase]) -> None:
    for i in range(len(phases) - 1):
        a, b = phases[i], phases[i + 1]
        if a.end_date >= b.start_date:
            raise IngestionError(
                f"Phases '{a.name}' and '{b.name}' overlap: "
                f"{a.name} ends {a.end_date}, {b.name} starts {b.start_date}"
            )


def _read_recurring(path: Path) -> list[RecurringCashFlow]:
    df = _read_sheet(path, "Recurring_Cash_Flows")
    _require_columns(
        df,
        ["Name", "Direction", "Amount", "Frequency", "Start_Date", "End_Date"],
        "Recurring_Cash_Flows",
    )

    flows = []
    for _, row in df.iterrows():
        name = str(row["Name"])
        direction = _parse_direction(str(row["Direction"]), name)
        frequency = _parse_frequency(str(row["Frequency"]), name)
        amount = float(row["Amount"])
        start = _to_date(row["Start_Date"])
        end = _to_date(row["End_Date"])

        try:
            flows.append(
                RecurringCashFlow(
                    name=name,
                    direction=direction,
                    amount=amount,
                    frequency=frequency,
                    start_date=start,
                    end_date=end,
                )
            )
        except ValueError as e:
            raise IngestionError(str(e)) from e

    return flows


def _read_actuals(path: Path) -> list[LiquidityActual]:
    df = _read_sheet(path, "Actuals")
    if df.empty:
        return []
    _require_columns(df, ["Date", "Liquidity"], "Actuals")
    actuals = []
    for _, row in df.iterrows():
        d = _to_date(row["Date"])
        if d is None:
            raise IngestionError("Actuals: Date is required for each row")
        try:
            amount = float(row["Liquidity"])
        except (ValueError, TypeError) as e:
            raise IngestionError(f"Actuals: invalid Liquidity value on {d}: {e}") from e
        actuals.append(LiquidityActual(date=d, amount=amount))
    actuals.sort(key=lambda a: a.date)
    return actuals


def _read_one_offs(path: Path) -> list[OneOffCashFlow]:
    df = _read_sheet(path, "One_Off_Cash_Flows")
    _require_columns(
        df,
        ["Name", "Direction", "Amount", "Date"],
        "One_Off_Cash_Flows",
    )

    flows = []
    for _, row in df.iterrows():
        name = str(row["Name"])
        direction = _parse_direction(str(row["Direction"]), name)
        amount = float(row["Amount"])
        d = _to_date(row["Date"])

        if d is None:
            raise IngestionError(f"Cash flow '{name}': Date is required for one-off")

        try:
            flows.append(
                OneOffCashFlow(
                    name=name,
                    direction=direction,
                    amount=amount,
                    date=d,
                )
            )
        except ValueError as e:
            raise IngestionError(str(e)) from e

    return flows

File: src/ingest.ts
/**
 * Parse raw spreadsheet rows into domain models. Ported from budgeteer/ingest.py.
 *
 * This layer is agnostic to where rows come from: a DataSource yields each sheet
 * as an array of rows (header row first), whether read from an uploaded ODS
 * (SheetJS) or, later, the Google Sheets API. Date cells must already be JS Date
 * objects — the DataSource owns any serial-number conversion.
 */
import { civilDate, compareDates, formatISO } from "./dates";
import {
  type AnyCashFlow,
  Direction,
  Frequency,
  IngestionError,
  LiquidityActual,
  OneOffCashFlow,
  Phase,
  RecurringCashFlow,
} from "./models";

export type Cell = string | number | Date | null | undefined;
/** A sheet as rows, with the header as the first row (SheetJS `header:1` shape). */
export type SheetRows = Cell[][];
export type SheetSet = Record<string, SheetRows>;

/** Supplies raw sheet rows. Implemented by OdsUploadSource / GoogleSheetsSource. */
export interface DataSource {
  load(): Promise<SheetSet>;
}

export interface ParsedInputs {
  phases: Phase[];
  cashFlows: AnyCashFlow[];
  actuals: LiquidityActual[];
}

export async function loadInputs(source: DataSource): Promise<ParsedInputs> {
  return parseInputs(await source.load());
}

export function parseInputs(sheets: SheetSet): ParsedInputs {
  const phases = readPhases(sheets);
  const recurring = readRecurring(sheets);
  const oneOffs = readOneOffs(sheets);
  const actuals = readActuals(sheets);
  return { phases, cashFlows: [...recurring, ...oneOffs], actuals };
}

function isEmpty(cell: Cell): boolean {
  return cell === null || cell === undefined || cell === "";
}

/**
 * Turn a sheet into column-keyed records, dropping all-empty rows and validating
 * required columns. Returns null records list when the sheet has no data rows.
 */
function readTable(sheets: SheetSet, name: string, required: string[]): Record<string, Cell>[] {
  const rows = sheets[name];
  if (rows === undefined) {
    throw new IngestionError(`Sheet '${name}' not found`);
  }
  const header = (rows[0] ?? []).map((c) => String(c ?? "").trim());
  const dataRows = rows.slice(1).filter((row) => !row.every(isEmpty));

  const missing = required.filter((col) => !header.includes(col)).sort();
  if (missing.length > 0) {
    throw new IngestionError(`Sheet '${name}': missing columns ${JSON.stringify(missing)}`);
  }

  return dataRows.map((row) => {
    const record: Record<string, Cell> = {};
    header.forEach((col, i) => {
      record[col] = row[i] ?? null;
    });
    return record;
  });
}

function toDate(val: Cell): Date | null {
  if (isEmpty(val)) {
    return null;
  }
  if (val instanceof Date) {
    if (Number.isNaN(val.getTime())) {
      return null;
    }
    return civilDate(val.getUTCFullYear(), val.getUTCMonth() + 1, val.getUTCDate());
  }
  const iso = String(val).slice(0, 10);
  const [y, m, d] = iso.split("-").map(Number);
  if (!y || !m || !d || Number.isNaN(y) || Number.isNaN(m) || Number.isNaN(d)) {
    return null;
  }
  return civilDate(y, m, d);
}

function titleCase(s: string): string {
  return s.toLowerCase().replace(/(^|\s)\S/g, (c) => c.toUpperCase());
}

function parseDirection(val: Cell, itemName: string): Direction {
  const candidate = titleCase(String(val ?? "").trim());
  if (candidate === Direction.Inflow || candidate === Direction.Outflow) {
    return candidate as Direction;
  }
  throw new IngestionError(
    `Cash flow '${itemName}': Direction must be 'Inflow' or 'Outflow', got '${val}'`
  );
}

function parseFrequency(val: Cell, itemName: string): Frequency {
  const candidate = titleCase(String(val ?? "").trim());
  if (candidate === Frequency.Monthly || candidate === Frequency.Annually) {
    return candidate as Frequency;
  }
  throw new IngestionError(
    `Cash flow '${itemName}': Frequency must be 'Monthly' or 'Annually', got '${val}'`
  );
}

function asIngestionError<T>(fn: () => T): T {
  try {
    return fn();
  } catch (e) {
    if (e instanceof IngestionError) {
      throw e;
    }
    throw new IngestionError(e instanceof Error ? e.message : String(e));
  }
}

function readPhases(sheets: SheetSet): Phase[] {
  const records = readTable(sheets, "Phases", ["Name", "Start_Date", "End_Date"]);
  const phases = records.map((row) => {
    const name = String(row.Name);
    const start = toDate(row.Start_Date);
    const end = toDate(row.End_Date);
    if (start === null || end === null) {
      throw new IngestionError(`Phase '${name}': Start_Date and End_Date are required`);
    }
    return asIngestionError(() => new Phase(name, start, end));
  });

  phases.sort((a, b) => compareDates(a.startDate, b.startDate));
  validateNoOverlaps(phases);
  return phases;
}

function validateNoOverlaps(phases: Phase[]): void {
  for (let i = 0; i < phases.length - 1; i++) {
    const a = phases[i] as Phase;
    const b = phases[i + 1] as Phase;
    if (a.endDate.getTime() >= b.startDate.getTime()) {
      throw new IngestionError(
        `Phases '${a.name}' and '${b.name}' overlap: ${a.name} ends ${formatISO(a.endDate)}, ${b.name} starts ${formatISO(b.startDate)}`
      );
    }
  }
}

function readRecurring(sheets: SheetSet): RecurringCashFlow[] {
  const records = readTable(sheets, "Recurring_Cash_Flows", [
    "Name",
    "Direction",
    "Amount",
    "Frequency",
    "Start_Date",
    "End_Date",
  ]);
  return records.map((row) => {
    const name = String(row.Name);
    const direction = parseDirection(row.Direction, name);
    const frequency = parseFrequency(row.Frequency, name);
    const amount = Number(row.Amount);
    const start = toDate(row.Start_Date);
    const end = toDate(row.End_Date);
    return asIngestionError(
      () => new RecurringCashFlow(name, direction, amount, frequency, start, end)
    );
  });
}

function readOneOffs(sheets: SheetSet): OneOffCashFlow[] {
  const records = readTable(sheets, "One_Off_Cash_Flows", ["Name", "Direction", "Amount", "Date"]);
  return records.map((row) => {
    const name = String(row.Name);
    const direction = parseDirection(row.Direction, name);
    const amount = Number(row.Amount);
    const d = toDate(row.Date);
    if (d === null) {
      throw new IngestionError(`Cash flow '${name}': Date is required for one-off`);
    }
    return asIngestionError(() => new OneOffCashFlow(name, direction, amount, d));
  });
}

function readActuals(sheets: SheetSet): LiquidityActual[] {
  if (sheets.Actuals === undefined) {
    throw new IngestionError("Sheet 'Actuals' not found");
  }
  const dataRows = sheets.Actuals.slice(1).filter((row) => !row.every(isEmpty));
  if (dataRows.length === 0) {
    return [];
  }

  const records = readTable(sheets, "Actuals", ["Date", "Liquidity"]);
  const actuals = records.map((row) => {
    const d = toDate(row.Date);
    if (d === null) {
      throw new IngestionError("Actuals: Date is required for each row");
    }
    const amount = Number(row.Liquidity);
    if (Number.isNaN(amount)) {
      throw new IngestionError(
        `Actuals: invalid Liquidity value on ${formatISO(d)}: ${row.Liquidity}`
      );
    }
    return new LiquidityActual(d, amount);
  });

  actuals.sort((a, b) => compareDates(a.date, b.date));
  return actuals;
}
